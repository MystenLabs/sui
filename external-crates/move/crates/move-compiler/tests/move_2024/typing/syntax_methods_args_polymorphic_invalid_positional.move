module a::invalid {

    public struct A<T> has drop { t: vector<T> }

    #[syntax(index)]
    public fun borrow_a<Q,T>(_s: &A<Q>, _j: T): &Q { abort 0 }

    #[syntax(index)]
    public fun borrow_mut_a<T,Q>(_s: &mut A<T>, _j: T): &mut T { abort 0 }

    public struct B<T> has drop { t: vector<T> }

    #[syntax(index)]
    public fun borrow_b<Q,T>(_s: &B<Q>, _j: Q): &T { abort 0 }

    #[syntax(index)]
    public fun borrow_mut_b<T,Q>(_s: &mut B<T>, _j: T): &mut T { abort 0 }

    public struct C<T> has drop { t: vector<T> }

    #[syntax(index)]
    public fun borrow_c<Q,T>(_s: &C<T>, _j: Q): &Q { abort 0 }

    #[syntax(index)]
    public fun borrow_mut_c<T,Q>(_s: &mut C<T>, _j: T): &mut T { abort 0 }

}
