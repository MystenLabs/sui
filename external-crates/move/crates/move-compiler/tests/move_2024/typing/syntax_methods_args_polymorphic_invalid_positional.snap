error[E04034]: 'syntax' method types differ
  ┌─ tests/move_2024/typing/syntax_methods_args_polymorphic_invalid_positional.move:6:45
  │
6 │     public fun borrow_a<Q,T>(_s: &A<Q>, _j: T): &Q { abort 0 }
  │                           -                 ^ This parameter has type 'T'
  │                           │                  
  │                           Type parameter T appears in position 2 here
  ·
9 │     public fun borrow_mut_a<T,Q>(_s: &mut A<T>, _j: T): &mut T { abort 0 }
  │                             -                       -
  │                             │                       │
  │                             │                       Expected this parameter to have type 'Q'
  │                             │                       It has type 'T'
  │                             Type parameter T appears in position 1 here
  │
  = Type parameters must be used the same by position, not name
  = Index operation non-subject parameter types must match exactly

error[E04034]: 'syntax' method types differ
   ┌─ tests/move_2024/typing/syntax_methods_args_polymorphic_invalid_positional.move:14:49
   │
14 │     public fun borrow_b<Q,T>(_s: &B<Q>, _j: Q): &T { abort 0 }
   │                           -                     ^^ This index function returns type '&T'
   │                           │                      
   │                           Type parameter T appears in position 2 here
   ·
17 │     public fun borrow_mut_b<T,Q>(_s: &mut B<T>, _j: T): &mut T { abort 0 }
   │                             -                           ------
   │                             │                           │
   │                             │                           Expected this mutable index function to return type '&mut Q'
   │                             │                           It returns type '&mut T'
   │                             Type parameter T appears in position 1 here
   │
   = Type parameters must be used the same by position, not name
   = These functions must return the same type, differing only by mutability

error[E04034]: 'syntax' method types differ
   ┌─ tests/move_2024/typing/syntax_methods_args_polymorphic_invalid_positional.move:22:34
   │
22 │     public fun borrow_c<Q,T>(_s: &C<T>, _j: Q): &Q { abort 0 }
   │                           -      ^^^^^ This index function subject has type '&a::invalid::C<T>'
   │                           │       
   │                           Type parameter T appears in position 2 here
   ·
25 │     public fun borrow_mut_c<T,Q>(_s: &mut C<T>, _j: T): &mut T { abort 0 }
   │                             -        ---------
   │                             │        │
   │                             │        Expected this mutable index function subject to have type '&mut a::invalid::C<Q>'
   │                             │        It has type '&mut a::invalid::C<T>'
   │                             Type parameter T appears in position 1 here
   │
   = Type parameters must be used the same by position, not name
   = These functions must take the same subject type, differing only by mutability

